### PR TITLE
network: change link state only when new address or route will be assigned

### LINKED ISSUE #18453: `test_ipv6_token_*` subtests (systemd-networkd-tests.py) intermittently fail
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Arch Linux & CentOS 7

Some of the `test_ipv6_token_*` subtests of `systemd-networkd-tests.py` started failing sporadically. According to CentOS CI logs the first fail happened around today's midnight (2021-02-03 12:50 AM CET).

```
======================================================================
ERROR: test_ipv6_token_prefixstable_without_address (__main__.NetworkdRATests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/test/test-network/systemd-networkd-tests.py", line 3519, in test_ipv6_token_prefixstable_without_address
    self.wait_online(['veth99:routable', 'veth-peer:degraded'])
  File "/build/test/test-network/systemd-networkd-tests.py", line 575, in wait_online
    check_output(*args, env=env)
  File "/build/test/test-network/systemd-networkd-tests.py", line 49, in check_output
    return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
  File "/usr/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/systemd-meson-build/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.

----------------------------------------------------------------------
```
Logs:
 * [Arch+ASan] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/2468/artifact//systemd-centos-ci/artifacts_all/artifacts_4_LY95/vagrant-logs.pIs/vagrant-arch-sanitizers-clang-testsuite.exh/systemd-networkd_sanitizers_FAIL.log

```
======================================================================
ERROR: test_ipv6_token_static (__main__.NetworkdRATests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/./test/test-network/systemd-networkd-tests.py", line 3497, in test_ipv6_token_static
    self.wait_online(['veth99:routable', 'veth-peer:degraded'])
  File "/build/./test/test-network/systemd-networkd-tests.py", line 575, in wait_online
    check_output(*args, env=env)
  File "/build/./test/test-network/systemd-networkd-tests.py", line 49, in check_output
    return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
  File "/usr/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.

----------------------------------------------------------------------
```
Logs:
 * [Arch] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/2489/artifact//systemd-centos-ci/artifacts_53cRlw/vagrant-logs.7GP/vagrant-arch-testsuite.P2P/systemd-networkd-tests.py_FAIL.log
 * [Arch+ASan] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/2471/artifact//systemd-centos-ci/artifacts_all/artifacts_zsq5YI/vagrant-logs.huX/vagrant-arch-sanitizers-clang-testsuite.T2c/systemd-networkd_sanitizers_FAIL.log
 * [Arch] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/2501/artifact//systemd-centos-ci/artifacts_RIEicG/vagrant-logs.naS/vagrant-arch-testsuite.w9K/systemd-networkd-tests.py_FAIL.log

```
======================================================================
ERROR: test_ipv6_token_prefixstable (__main__.NetworkdRATests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/test/test-network/systemd-networkd-tests.py", line 3509, in test_ipv6_token_prefixstable
    self.wait_online(['veth99:routable', 'veth-peer:degraded'])
  File "/build/test/test-network/systemd-networkd-tests.py", line 575, in wait_online
    check_output(*args, env=env)
  File "/build/test/test-network/systemd-networkd-tests.py", line 49, in check_output
    return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
  File "/usr/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/systemd-meson-build/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--interface=veth-peer:degraded']' returned non-zero exit status 1.

----------------------------------------------------------------------
```
Logs:
 * [Arch+ASan] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/2476/artifact//systemd-centos-ci/artifacts_all/artifacts_XtjQNr/vagrant-logs.dId/vagrant-arch-sanitizers-clang-testsuite.0Ms/systemd-networkd_sanitizers_FAIL.log

```
======================================================================
FAIL: test_ipv6_prefix_delegation (__main__.NetworkdRATests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 3478, in test_ipv6_prefix_delegation
    self.wait_online(['veth99:routable', 'veth-peer:degraded'])
  File "./test/test-network/systemd-networkd-tests.py", line 583, in wait_online
    self.wait_operstate(link.split(':')[0], None, setup_state, setup_timeout)
  File "./test/test-network/systemd-networkd-tests.py", line 548, in wait_operstate
    self.fail(f'Timed out waiting for {link} to reach state {operstate}/{setup_state}')
AssertionError: Timed out waiting for veth99 to reach state \S+/configured

----------------------------------------------------------------------
```
Logs:
 * [CentOS 7] https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-centos7/2470/artifact//systemd-centos-ci/artifacts_LAsfpE/testsuite-logs-upstream.8dd/systemd-networkd-tests.py_FAIL.log

/cc @bluca, @yuwata

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #18453.

### CODE REVIEW

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -326,7 +326,7 @@ static int address_add_foreign(Link *link, const Address *in, Address **ret) {
 
 static int address_add(Link *link, const Address *in, Address **ret) {
         Address *address;
-        int r;
+        int r, k;
```
Comment (poettering) [inline src/network/networkd-address.c]: doesn't matter, but why not `bool is_new;` here?

i mean, bools neatly convert to ints, so if you do `return is_new` at the end, this still behaves exactly the same, but is just ~1.2 times prettier ;-)

File: src/network/networkd-route.c [hunk 2]
```diff
@@ -1055,6 +1055,7 @@ int route_configure(
 
         _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *req = NULL;
         int r, k = 0;
+        Route *nr;
```
Comment (poettering) [inline src/network/networkd-route.c:1080]: it the `else` branch further down is entered, i.e. the one where `route_add_and_setup_time()` is false, then `nr` won't be initialized, and you return rubbish to the caller. either initialize the variable to NULL here up top, or in that `else` branch

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/networkd-address.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-address.c]: Done.

File: src/network/networkd-route.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-route.c:1080]: Hm? `route_add_and_setup_time()` sets `nr` when it returns non-negative, i.e. 0 or 1. So, no rubbish is returned to the caller. Or do I misunderstand your point?

Comment (keszybz) [reply src/network/networkd-route.c:1080]: Lennart meant the conditional with `ordered_set_isempty(route->multipath_routes)`. In the `else` branch, `nr` is not set.

File: src/network/networkd-dhcp6.c [hunk 3]
```diff
@@ -137,7 +137,7 @@ static int dhcp6_pd_remove_old(Link *link, bool force) {
         assert(link);
         assert(link->manager);
 
-        if (!force && (link->dhcp6_pd_address_messages != 0 || link->dhcp6_pd_route_configured != 0))
+        if (!force && (link->dhcp6_pd_address_messages != 0 || link->dhcp6_pd_route_messages != 0))
```
Comment (keszybz) [inline src/network/networkd-dhcp6.c:140]: "fix typo" doesn't really do just to this patch ;) I'd say "fix condition check" instead.

**Review by keszybz — CHANGES_REQUESTED** [review]

File: src/network/networkd-route.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-route.c:1080]: But in the `else` branch we have `assert(!ret);`. So, uninitialized `nr` is no problem. No?

File: src/network/networkd-dhcp6.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-dhcp6.c:140]: The commit message is updated.

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**